// ng-gapi/lib/GoogleApiModule.js
import { GoogleApiService } from "./GoogleApiService";
import { NgModule } from "@angular/core";
import { GoogleAuthService } from "./GoogleAuthService";
var GoogleApiModule = (function () {
    function GoogleApiModule() {
    }
    GoogleApiModule.setConfig = function (apiConfig) {
        return {
            ngModule: GoogleApiModule,
            providers: [
                {
                    provide: GoogleApiService,
                    useFactory:function(){ return GoogleApiService.factory(apiConfig) },
                },
                {
                    provide: GoogleAuthService,
                    useFactory: GoogleAuthService.factory,
                    deps: [
                        GoogleApiService
                    ]
                }
            ]
        };
    };
    return GoogleApiModule;
}());
export { GoogleApiModule };
GoogleApiModule.decorators = [
    { type: NgModule },
];
/** @nocollapse */
GoogleApiModule.ctorParameters = function () { return []; };
